### PR TITLE
Check write permission before creating the DP keys directory.

### LINKED ISSUE #2293: AddDataProtection Beta1 -> Beta2
On migrating from Beta1 to Beta2 I see that data protection was added.

This has caused our application not to startup  
``` csharp
"Access to the path 'c:\\**\\**\\**\\App_Data\\**\\***\\DataProtection-Keys' is denied."
```

I tried adding redis

``` csharp
services.AddDataProtection().PersistKeysToRedis 
```

but no luck

If there was a way to set The type of DataProtection

As well as  have the ability to disable it since we use Orchard on readonly apis

### PR DESCRIPTION
**Pull request by jtkech**
Try to fix #2293.

### CODE REVIEW

File: src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -182,20 +183,34 @@ private static void AddDataProtection(OrchardCoreBuilder builder)
...
+
+                try
+                {
+                    directory = Directory.CreateDirectory(Path.Combine(
+                        options.Value.ShellsApplicationDataPath,
+                        options.Value.ShellsContainerName,
+                        settings.Name, "DataProtection-Keys"));
+                }
+                catch (UnauthorizedAccessException)
+                {
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs]: Log.Error

File: src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -182,20 +183,34 @@ private static void AddDataProtection(OrchardCoreBuilder builder)
...
+                var dpBuilder = new ServiceCollection()
                     .AddDataProtection()
                     .PersistKeysToFileSystem(directory)
-                    .SetApplicationName(settings.Name)
-                    .Services;
+                    .SetApplicationName(settings.Name);
+
+                if (directory != null)
+                {
+                    dpBuilder.PersistKeysToFileSystem(directory);
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs]: The question is where is the thing stored? Also will it still be isolated by tenant?
I would have preferred a way to configure the location of the keys instead.

File: src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -182,20 +183,34 @@ private static void AddDataProtection(OrchardCoreBuilder builder)
...
+                }
 
                 // Re-register the data protection services to be tenant-aware so that modules that internally
                 // rely on IDataProtector/IDataProtectionProvider automatically get an isolated instance that
                 // manages its own key ring and doesn't allow decrypting payloads encrypted by another tenant.
                 // By default, the key ring is stored in the tenant directory of the configured App_Data path.
-                var collection = new ServiceCollection()
+                var dpBuilder = new ServiceCollection()
                     .AddDataProtection()
                     .PersistKeysToFileSystem(directory)
```
[the remark this answers is not in the document]
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs]: Oups.

File: src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs]: Okay.

File: src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore/OrchardCore.Modules/Extensions/ServiceCollectionExtensions.cs]: It depends on the environment, see here

https://docs.microsoft.com/en-us/aspnet/core/security/data-protection/configuration/default-settings?view=aspnetcore-2.1

Where they say

> The first three options above should provide good defaults

And also

> If none of these conditions match, keys aren't persisted outside of the current process. When the process shuts down, all generated keys are lost.

Maybe need to be confirmed by @PinpointTownes

**Comment by jtkech** [comment]
We could also not create a directory and not call `PersistKeysToFileSystem()` in `AddDataProtection()`. Then do an helper to do this, and that would be called in `AddOrchardCms()` but not in `AddOrchardCore()`.

**Comment by PinpointTownes** [comment]
> Maybe need to be confirmed by @PinpointTownes

Yup, if you don't set the directory path yourself (via calling `PersistKeysToFileSystem`), the default Data Protection storage resolution logic will automatically kick in. This is terrible in our case as it means we could end up registering a shared storage like `%LOCALAPPDATA%` or the registry for all tenants. It's only when both the `%LOCALAPPDATA%` folder and the registry are unavailable that DP falls back to the in-memory repository.

Having a shared key ring would make tenants terribly vulnerable, as things like cookies/tokens or encrypted settings created for tenant A could be read by tenant B (and vice versa).

That being said, I realize the current code is not ideal as it completely prevents a host from starting if the tenant-specific DP folder can't be created. To improve that, I suggest 2 options:

  - Creating our own `EphemeralXmlRepository` (the one in ASP.NET Core DP is internal and we have no way to register it, except by using reflection, which sucks) and falling back to it when we can't create the directory. This way, we completely bypass the default DP logic that may result in a shared repository being used.

  - Falling back to the host-provided DP services **and** registering a tenant-aware `IDataProtectionProvider` instance at the tenant level to provide the isolation between tenants. This way, we use DP's default logic resolution (so a shared place may be selected), but we still provide the necessary isolation at a different layer. This was our initial plan before we moved to a per-tenant-DP-keys model per @sebastienros' demand (not for security, just to make exporting keys easier): https://github.com/OrchardCMS/OrchardCore/issues/1391.

**Comment by jtkech** [comment]
Okay, I will wait for the chosen option.

So, for the 1st option we would have to create our own `EphemeralXmlRepository`.

And for the 2nd option, if we can't create the directory, we just do nothing. Right?.

**Comment by jtkech** [comment]
Hmm, maybe the 2nd option is what i'm doing here. If we can't create the directory we still call `AddDataProtection()` at the tenant level so with its default logic but isolated, we just still use `.SetApplicationName(settings.Name)`. But not sure ;)

**Comment by PinpointTownes** [comment]
> Hmm, maybe the 2nd option is what i'm doing here. If we can't create the directory we still call AddDataProtection() at the tenant level so with its default logic but isolated, we just still use .SetApplicationName(settings.Name). But not sure ;)

Oh yeah, nice catch! I missed we're explicitly overriding the discriminator to be per-tenant (hem... I wrote this code, which proves I have terrible memory :smile:)

In this case, the PR as-is should be fine.

**Comment by PinpointTownes** [comment]
> Hmm, maybe the 2nd option is what i'm doing here.

Not exactly: with your PR (and more generally with the existing implementation in beta2), if we register a different repository (e.g Redis) at the host-level, it won't be used by the tenant DP services.

**Comment by jtkech** [comment]
Cool 👍

**Comment by jtkech** [comment]
Hmm, but per tenant we can still cange the storage as the `OrchardCore.DataProtection.Azure` is doing in its startup which is run after because of its `Order` property.

    services.AddDataProtection().PersistKeysToAzureBlobStorage(GetBlobContainer(), blobName);

Here the tryadd pattern is ok because it acts on the already registered services at the tenant level, and we already have done some cleanups (e.g duplicate startup filters).

**Comment by jtkech** [comment]
>  if we register a different repository (e.g Redis) at the host-level, it won't be used by the tenant DP services

Hmm, this is already the case but maybe you mean just in case we can't create a directory.

> Sure, but you have to write a module for that

We can do that from the host with our helper but maybe you mean to be able to enable it per tenant.

In fact i'm lost ;) but we can do everything. Maybe you mean that if we can't create a directory we would have to retrieve the kind of repository that may have been registered at the host level. If so i could do that.

Otherwise, if you think that creating an `EphemeralXmlRepository` is better.

**Comment by PinpointTownes** [comment]
> Hmm, this is already the case but maybe you mean just in case we can't create a directory.

Yeah. I'm not saying we should absolutely fall back to the host-level DP, just that we may consider it if we think it's desirable.

I guess we could also make that configurable, by allowing the users to opt for per-tenant DP keys or shared keys stored in the host DP repository. @sebastienros any remark?

**Comment by jtkech** [comment]
So, here i suggest that on a read only system we use by default an `EphemeralXmlRepository` as the first option proposed by @PinpointTownes. Knowing that a module or the application can override this option, but just to not fail on `CreateDirectory()` and still have a non null key repository.

I tried it, it works without the `DataProtection-Keys` folder and i can login in the admin. But when you stop and restart the application you have to login again to go to the admin because keys are stored in memory.

**Comment by sebastienros** [comment]
But here the issue is that it was not writable in prod, and people won't enable the App_data one by themselves, so it needs to be a default. IMO there is currently nothing to fix. Or if you know you can't have write access to the deployed folder then you need to opt-in for another store. The same way if you know you will scale up you will need shared services like Redis, Blog Storage, ... for all Orchard components. I'd suggest as a solution to at least document a way to enable such a shared storage for keys.

**Comment by sebastienros** [comment]
I think the only change to do is to display a better error message (like you did) when the folder can't be created. It should show the folder name and say it should have write access, or that another data protection store should be used.

**Comment by jtkech** [comment]
Oh yes, i think you meant to make it configurable through a config file and then not try to create the directory if the option is not the file system. Okay cool, i close this one.

**Pull request closed without merging**